chatsky/core/vector_store.py

```diff
@@ -0,0 +1,188 @@
+    model_name: str = "DeepPavlov/rubert-base-cased"
+    """Name of the embedding model. Defaults to "DeepPavlov/rubert-base-cased"."""
+
+
+class VectorStoreService:
+    """
+    Abstract base class for vector store services.
+
+    Provides common interface for document storage and retrieval operations.
+    Must be subclassed with specific vector database implementations.
+    """
+
+    def __init__(self, config: VectorStoreConfig):
+        """
+        Initialize the vector store service with configuration.
+
+        :param config: Configuration parameters for the vector store.
+        """
+        self.config = config
+        self.initialized = False
+
+    async def initialize(self, ctx: Context):
+        """
+        Initialize connection to the vector database and register with context.
+
+        :param ctx: Framework context object where the vector store will be registered.
+
+        :raises RuntimeError: If initialization fails.
+        """
+        if not self.initialized:
+            ctx.framework_data.vector_store = self
+            self.initialized = True
+            logger.info("Vector store initialized")
+
+    async def add_documents(self, documents: List[Message], metadata: dict[Optional]):
+        """
+        Add documents to the vector store.
+
+        :param documents: List of Message objects to be stored.
+        :param metadata: Optional metadata to associate with documents.
+
+        :raises NotImplementedError: Must be implemented by subclasses.
+        """
+        raise NotImplementedError
+
+    async def retrieve(self, query: str) -> List[Message]:
+        """
+        Retrieve documents similar to the query.
+
+        :param query: Search query string.
+
+        :return: List of matching Message objects.
+
+        :raises NotImplementedError: Must be implemented by subclasses.
+        """
+        raise NotImplementedError
+
+
+class ChromaVectorStore(VectorStoreService):
+    """
+    Chroma DB implementation of VectorStoreService.
+
+    Provides concrete implementation using Chroma vector database.
+    """
+
+    def __init__(self, config: VectorStoreConfig):
+        """
+        Initialize Chroma client with given configuration.
+
+        :param config: Configuration parameters.
+
+        :raises Exception: If initialization fails.
+        """
+        super().__init__(config)
+        try:
+            self.embeddings = HuggingFaceEmbeddings(model_name=config.model_name)
+            self.client = Chroma(
+                collection_name=config.collection_name,
+                embedding_function=self.embeddings,
+                persist_directory=config.persist_directory,
+            )
+            logger.info(
+                "Chroma client initialized for collection: %s", config.collection_name
+            )
+        except Exception as e:
+            logger.error("Failed to initialize Chroma: %s", str(e))
+            raise
```

Comment:
I don't like the error log since it would lead to log duplication after the re-raised exception is handled again.